feat(docs): add keystore monitoring section in delegated stake (#17810)
AztecProtocol/aztec-packages: docs/docs/the_aztec_network/operation/sequencer_management/running_delegated_stake.md

- Foundry installed for `cast` commands

## How Delegated Stake Works

You register with the StakingRegistry contract and add sequencer identities (keystores) to a queue. When delegators stake to your provider, the system:

1. Dequeues one keystore from your provider queue
2. Creates a [Split contract](https://docs.splits.org/core/split) for reward distribution
3. Registers the sequencer into the staking queue using the dequeued keystore

### Reward Distribution

When a delegator stakes to your provider, a Split contract is automatically created to manage reward distribution. You configure your sequencer to use the Split contract address as the coinbase (see [After Delegation: Configure Sequencer Coinbase](#after-delegation-configure-sequencer-coinbase)).

Rewards are distributed according to your agreed commission rate:
- **Provider commission**: Your `providerRewardsRecipient` address receives your commission rate (e.g., 5% for 500 basis points)
- **Delegator rewards**: The delegator's Aztec Token Vault (ATV) receives the remaining percentage

**Rewards flow:**
1. Rewards accumulate in the rollup under the coinbase address (the Split contract)

> 2. After governance unlocks rewards, anyone can release them from the rollup to the coinbase

Reviewer: ```suggestion
2. After governance unlocks rewards, anyone can release them from the rollup to the `coinbase` address. 
```

- Each delegation creates a unique Split contract—configure each sequencer with its specific Split contract address
- Restart your sequencer node after updating the keystore for changes to take effect

## Monitoring Keystore Availability

As a provider, you must maintain sufficient sequencer identities (keystores) in your queue to handle incoming delegations. When a delegator stakes to your provider and your queue is empty, they cannot activate a sequencer—this results in a poor delegator experience and lost opportunity.

### Why Monitoring Matters

Each time a delegator stakes to your provider:
1. One keystore is dequeued from your provider queue
2. A sequencer is activated using that keystore
3. Your available keystore count decreases by one

If your queue runs empty, new delegations cannot activate sequencers until you add more keystores. This could cause delegators to choose other providers.

### Checking Available Keystores

Check your current keystore queue with this call:

```bash
# Check provider queue
cast call [STAKING_REGISTRY_ADDRESS] \
  "providerQueues(uint256)" \
  [YOUR_PROVIDER_IDENTIFIER] \
  --rpc-url [RPC_URL]

> ```

Reviewer: ```suggestion
```bash
# Check provider queue
cast call [STAKING_REGISTRY_ADDRESS] \
  "getProviderQueueLength(uint256)" \
  [YOUR_PROVIDER_IDENTIFIER] \
  --rpc-url [RPC_URL]
```

Reviewer: This returns only the number of available keystores which is what you seem to be trying to do here.

Author: Good call, thanks for the 👁️. Addressed !

# Calculate queue length (last - first)
QUEUE_LENGTH=$((LAST - FIRST))

echo "Queue length: $QUEUE_LENGTH (first: $FIRST, last: $LAST)"

# Check if queue is running low
if [ "$QUEUE_LENGTH" -lt "$THRESHOLD" ]; then
  echo "WARNING: Keystore queue running low! Only $QUEUE_LENGTH keystores remaining."

  # Send alert (uncomment and configure your preferred method)
  # Slack/Discord webhook:
  # curl -X POST "$WEBHOOK_URL" -H "Content-Type: application/json" \
  #   -d "{\"text\":\"⚠️ Keystore queue low: $QUEUE_LENGTH remaining (threshold: $THRESHOLD)\"}"

  # Email via mail command:
  # echo "Keystore queue has $QUEUE_LENGTH keys remaining" | mail -s "Low Keystore Alert" your-email@example.com
fi
```

Make the script executable and schedule it with cron:

```bash
# Make the script executable
chmod +x /path/to/check-keystores.sh

# Edit crontab
crontab -e

# Add this line to check every 4 hours
0 */4 * * * /path/to/check-keystores.sh >> /var/log/keystore-monitor.log 2>&1
```

### Recommended Thresholds

- **Minimum queue length**: Never let your queue drop below 3-5 keystores
- **Replenishment trigger**: Add new keystores when the queue falls below 10

> - **Target queue size**: Maintain 20-50 keystores depending on your expected delegation volume

Reviewer: Suggest dropping the numbers here as they're sort of arbitrary

Author: Done !

- Each delegation creates a unique Split contract—configure each sequencer with its specific Split contract address
- Restart your sequencer node after updating the keystore for changes to take effect

## Monitoring Keystore Availability

As a provider, you must maintain sufficient sequencer identities (keystores) in your queue to handle incoming delegations. When a delegator stakes to your provider and your queue is empty, they cannot activate a sequencer—this results in a poor delegator experience and lost opportunity.

### Why Monitoring Matters

Each time a delegator stakes to your provider:
1. One keystore is dequeued from your provider queue
2. A sequencer is activated using that keystore
3. Your available keystore count decreases by one

If your queue runs empty, new delegations cannot activate sequencers until you add more keystores. This could cause delegators to choose other providers.

### Checking Available Keystores

Check your current keystore queue with this call:

```bash
# Check provider queue length
cast call [STAKING_REGISTRY_ADDRESS] \
  "getProviderQueueLength(uint256)" \
  [YOUR_PROVIDER_IDENTIFIER] \
  --rpc-url [RPC_URL]

> ```

Reviewer: ```suggestion
```bash
# Check provider queue length
cast call [STAKING_REGISTRY_ADDRESS] \
  "getProviderQueueLength(uint256) (uint256)" \
  [YOUR_PROVIDER_IDENTIFIER] \
  --rpc-url [RPC_URL]
```

Reviewer: just a style change, this returns the actual number instead of the hex value

  --private-key [ADMIN_PRIVATE_KEY]
```

:::note
Rate changes only apply to new delegations. Existing delegations retain the original commission rate they agreed to.
:::

## Verification

Verify your setup is working correctly.

### Check Provider Registration

Query the StakingRegistry to confirm your provider details:

> ```bash
> cast call [STAKING_REGISTRY_ADDRESS] \
>   "providerConfigurations(uint256)" \
>   [YOUR_PROVIDER_IDENTIFIER] \
>   --rpc-url [RPC_URL]
> ```

Reviewer: ```suggestion
```bash
cast call [STAKING_REGISTRY_ADDRESS] \
  "getProvider(uint256)" \
  "providerConfigurations(uint256) (address,uint16,address)" \
  [YOUR_PROVIDER_IDENTIFIER] \
  --rpc-url [RPC_URL]
```

Author: Suggestion was a bit wonky but I think I understood your point, have edited. Thanks !

> This returns your provider configuration including admin address, commission rate, and rewards recipient.

Reviewer: ```suggestion
This returns: 
1. The provider's admin address
2. The provider's commission rate in bps
3. The provider's rewards recipient 
```

### Verify Sequencer Identities

Check the keystores in your provider queue:

```bash
# Check provider queue
cast call [STAKING_REGISTRY_ADDRESS] \

>   "providerQueues(uint256)" \

Reviewer: I don't think this will work honestly. Have you tried this?

Author: Ah I thought it would return the actual items, have reverted these changes and explained it better.

## Best Practices

**Maintain Sufficient Keystores**: Set up automated monitoring to ensure your keystore queue never runs empty. See [Monitoring Keystore Availability](#monitoring-keystore-availability) for guidance on implementing alerts and recommended thresholds.

**Communicate Changes**: Inform delegators about commission rate changes, planned maintenance, or infrastructure updates. Good communication builds trust.

**Monitor Performance**: Track your sequencers' attestation rates, block proposals, and uptime. Poor performance may cause delegators to withdraw.

**Secure Your Keys**: The `providerAdmin` key controls your provider configuration. Store it securely and consider using a hardware wallet or multisig.

> **Set Competitive Commission Rates**: Balance profitability with competitiveness. Rates that are too high drive delegators to other providers; rates that are too low make operations unsustainable.

Reviewer: Why is the Foundation providing this guidance?

Author: Have removed this bulletpoint, it was just something extra and unnecessary
